RotationGesture + fixed scaleX, scaleY, rotationRadians going out of sync with layer.transform

saniul opened the pull request:
Implemented a rotation gesture as well as proposed changes discussed in #3

saniul commented:
Not sure how useful the centroid path tracking is for the rotation gesture if we’re not recording the rotation itself over time. Didn’t do anything special here to keep it simple.

NachoSoto commented on Prototope/Layer.swift:
```diff
@@ -253,13 +258,23 @@ public class Layer: Equatable, Hashable {
-		didSet { updateTransform() }
+	public var scaleX: Double {
+        get {
+            return layer.valueForKeyPath("transform.scale.x") as Double
```
I wonder if getting these values from the presentation layer would be more what we want? :)

saniul replied:
Thanks for having a look at this!

I feel like that’d be an implementation detail that would easily leak when using this API. Say you set the rotation (by setting it at the `layer.transform` level) and then immediately query the rotation (this would extract it from `layer.presentationLayer.transform`) - you will get the old value.

Besides, it feels like the whole point of this was to abstract away some of the UIKit, CA framework weirdness behind an API that’s simpler, quicker to use and understand for a limited set of use cases. Exposing the CA implementation detail of layer/presentationLayer duality seems to go against this goal (to me). But let me know if I’m wrong :)

NachoSoto replied:
I totally see what you mean, unfortunately I'm not sure we can hide UIKit completely. A parallel example to yours:
Say you animate a layer's rotation, and sometime during the animation you query the value, hoping to get the *current* rotation. You'd get the final value.

I'm not sure how we can effectively provide both values, while keeping the API simple and not exposing such CA details :/ But I haven't thought about this enough.

andymatuschak replied:
@saniul: You've totally got it right. I haven't actually written this out, but I intend to deal with this dichotomy through the distinction between `layer.animators.x.target` and `layer.x`. i.e. the first is "like the model value" and the second is "like the presentation value."

Yeah, we still have this semi-vestigial `Layer.animateWithDuration` method which doesn't use the POP animators, but I really don't like that API anyway. Gotta merge it with the dynamic animators as a convenience or otherwise make it go through the same paths. Haven't thought it through.

andymatuschak replied:
(@nsfmc, does my comment address your comment too?)

nsfmc replied:
:+1: @andymatuschak

nsfmc commented:
as i understand it, i feel like the more natural approach would be to do something along the lines of allowing folks to add something like `Animator.currentAnimationHandler` where the `animationHandler` would receive the layer and the current value of the animated property? i would expect that *only* this sort of handler would have the 'live' value of the animation at any given moment and that any other queryable values might be stale or optimistic/future values. Similar to what's going on with Heartbeat.swift where you can attach a handler for each frame and modify properties then.

i dunno, does that make any sense?

saniul commented:
Having a handler that can fire on every frame of the animation sounds useful and is definitely something that we should consider adding regardless of this issue.

However, it still won’t solve the fact that users will have to *think* about whether they want to get the underlying model value (query the layer’s properties) or the “inflight” value (check if there’s an animation and if so - attach an animation handler).

Note: Am I wrong in thinking that this is only an issue with CoreAnimation’s animations? AFAIK, POP animates by setting the model value on every display frame, so they will reflect what’s presented on the screen.

Anyhow, this discussion warrants a separate Issue of its own. And I feel like the eventual solution should stem from the envisioned use case, as well as whether or not Prototope will continue to support classic UIView/CAAnimations, as opposed to relying exclusively on POP animations.

andymatuschak commented on Prototope/Gesture.swift:
```diff
@@ -268,6 +268,79 @@ public class PanGesture: GestureType {
 	}
 }
 
+/** A rotation gesture recognizes a standard iOS rotation: it doesn't begin until the user's moved some number of points, then it tracks new touches coming and going over time as well as rotation relative to the beginning of the gesture and the current rotation velocity. It exposes simple access to the path of the center of all the touches and the rotation and rotation velocity values. */
```
s/moved some number of points/rotated by some number of degrees

andymatuschak commented on Prototope/Gesture.swift:
```diff
@@ -268,6 +268,79 @@ public class PanGesture: GestureType {
+    /** The handler will be invoked as the gesture recognizes and updates; it's passed the gesture's current
+    phase (see ContinuousGesturePhase documentation), rotation relative to the state at the beginning 
+    of the gesture in radians, rotation velocity in radians per second and also a touch sequence representing the center of all the touches involved in the rotation gesture. */
+    public convenience init(_ handler: (phase: ContinuousGesturePhase, rotationRadians: Double, rotationVelocity: Double, centroidSequence: TouchSequence<Int>) -> ()) {
```
Wish there were some good way to make defining this closure handler more friendly. This many positional arguments is rough!

One idea: what if we defined a `RotationGesture.State` `struct` subtype (with `phase`, `rotationRadians`, etc fields) and passed one of those? Then you could do lightweight things like this:

```
layer.gestures.append(RotationGesture { layer.rotationRadians = $0.rotationRadians })
```

andymatuschak commented on Prototope/Gesture.swift:
```diff
@@ -268,6 +268,79 @@ public class PanGesture: GestureType {
+            let rotationGesture = gestureRecognizer as UIRotationGestureRecognizer
+            switch rotationGesture.state {
+            case .Began:
+                // Reset the gesture to record rotation relative to the starting rotation
```
Hm, but the following line doesn't actually reset the gesture if I understand correctly.

saniul replied:
Yeah, I removed that comment in https://github.com/saniul/Prototope/commit/cedb32fe6dca19c82562ee4bebcf3e040f85ec91

andymatuschak commented:
Really cool that you did this, @saniul! Appreciate the contribution. :100: 

It seems like this API really wants to be able to emit a "rotation state sequence," in the same sense that the other APIs emit e.g. a touch sequence. I know from experience that I'd want to be able to refer to e.g. the current rotation radians vs. the previous rotation radians vs. the initial rotation radians—really, for the same reasons that you'd want to be able to do that with touch locations. What do you think?

saniul commented:
Totally agree, I didn’t implement recording the state sequence to keep things simple, but it does make a lot of sense.

One thing though, I can’t really reconcile [simplifying the handler closure’s arguments](https://github.com/Khan/Prototope/pull/4#discussion_r23699786) down to a single state `struct` with emitting a rotation state sequence. Looks like it’ll have to be something like `(ContinuousGesturePhase, GestureStateSequence<RotationGesture.State>, TouchSequence<Int>) -> ()`, unless we somehow unify `TouchSequence` and `GestureStateSequence` (which doesn’t seem right to me).

I’ll work on the changes later today.

andymatuschak commented:
What if:

```swift
struct SampleSequence<Sample> {
 ...
}

typealias TouchSequence = SampleSequence<TouchSample>

struct RotationSample {
  let rotationRadians: Double
  let centroid: TouchSample
  ...
}

typealias RotationSampleSequence = SampleSequence<RotationSample>
```

etc

saniul commented:
You’ll have to excuse my wariness for changing the existing TouchSequence API :) Yeah, seems like that is the way to go. I’ll have to extract the `currentVelocityInLayer` into function since extensions on specialized generic types aren’t allowed (related: https://twitter.com/indragie/status/560539062398750722)

andymatuschak commented:
Ooh, yeah. Hm hm. Maybe make TouchSequence a decorator on SampleSequence instead of making it a typealias, so you can keep that as a method?

> On Jan 28, 2015, at 2:46 PM, Saniul Ahmed <notifications@github.com> wrote:
> 
> You’ll have to excuse my wariness for changing the existing TouchSequence API :) Yeah, seems like that is the way to go. I’ll have to extract the currentVelocityInLayer into function since extensions on specialized generic types aren’t allowed (related: https://twitter.com/indragie/status/560539062398750722 <https://twitter.com/indragie/status/560539062398750722>)
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/Khan/Prototope/pull/4#issuecomment-71933037>.
>

saniul commented:
@andymatuschak How’s that https://github.com/saniul/Prototope/commit/4efa8ab01ee6917f4acf6e670ce38a5cc7c99172? Not sure if [this](https://github.com/saniul/Prototope/commit/4efa8ab01ee6917f4acf6e670ce38a5cc7c99172#diff-cb323b36c43bdd9204ab393c98b3ce44R418) is quite what you wanted me to do for TouchSequence.

saniul commented:
Any updates on this? Would you like me to implement it differently?

andymatuschak commented:
Very sorry for the delay—I got sidetracked by many other things! You've done an excellent job with that refactoring! Thanks so much.

saniul commented:
@andymatuschak No worries!

- [x] Added `rotationDegrees` and `velocityDegrees` to `RotationSample` and modified the `description` to `"<RotationSample: ⟳\(rotationDegrees)° ∂⟳\(velocityDegrees)°/s, \(rotationRadians)rad \(velocityRadians)rad/s, @\(centroid)>”`
- [x] Changed `SampleSequence`and `TouchSequence` internals to refer to the original typealias names to make the code more readable

saniul commented:
:boom: 

I’ll update the OhaiPrototype PR to point to the updated submodule in about an hour.

andymatuschak commented:
Woo! Thanks so much @saniul! Really appreciate the contribution :100:

nsfmc commented:
:heart:
